eox_nelp/api_clients/__init__.py

```diff
@@ -127,3 +96,98 @@ def make_get(self, path, payload):
             "error": True,
             "message": f"Invalid response with status {response.status_code}"
         }
+
+
+class AbstractOauth2ApiClient(AbstractApiClient):
+    """Abstract class for an OAuth 2.0 authentication API client.
+
+    This class provides basic functionality for an API client that requires
+    OAuth 2.0 authentication using the client ID and client secret.
+
+    Attributes:
+        client_id (str): Client ID for OAuth 2.0 authentication.
+        client_secret (str): Client secret for OAuth 2.0 authentication.
+    """
+
+    def __init__(self):
+        """Initialize an instance of AbstractOauth2ApiClient.
+
+        Gets the client ID and client secret from Django settings and then
+        calls the constructor of the parent class (AbstractApiClient).
+        """
+        self.client_id = getattr(settings, "FUTUREX_API_CLIENT_ID")
+        self.client_secret = getattr(settings, "FUTUREX_API_CLIENT_SECRET")
+
+        super().__init__()
+
+    def _authenticate(self):
+        """Authenticate the session with OAuth 2.0 credentials.
+
+        This method uses OAuth 2.0 client credentials (client ID and client secret)
+        to obtain an access token from the OAuth token endpoint. The access token
+        is then used to create and configure a requests session.
+
+        The access token is cached to minimize token requests to the OAuth server.
+
+        Returns:
+            requests.Session: Session authenticated with OAuth 2.0 credentials.
+        """
+        key = f"{self.client_id}-{self.client_secret}"
+        headers = cache.get(key)
+
+        if not headers:
+            client = BackendApplicationClient(client_id=self.client_id)
+            oauth = OAuth2Session(client_id=self.client_id, client=client)
+            authenticate_url = f"{self.base_url}/oauth/token"
+            response = oauth.fetch_token(
+                token_url=authenticate_url,
+                client_secret=self.client_secret,
+                include_client_id=True,
+            )
+            headers = {
+                "Authorization": f"{response.get('token_type')} {response.get('access_token')}"
+            }
+
+            cache.set(key, headers, response.get("expires_in", 300))
```

Comment:
Cache would only work for oauth2 and not basic auth?

Reply:
yes, the cache is just for oauth2  and the main reason is that both systems work different, both systems  send the authentication data in the `Authorization` header however the authentication data is generated in a different way, oauth2 uses a client_id and a client_secret to make a request to generate a token, so the backend is on charge of the generation, on the other hand, basic  authorization doesn't need to make a request since the authorization data  is just the `username:password`  in base64